File: packages/android/richtext-android/src/main/java/com/company/richtext/strip/Strip.kt
package com.company.richtext.strip

import com.company.richtext.ast.BlockNode
import com.company.richtext.ast.DocumentNode
import com.company.richtext.ast.InlineNode
import com.company.richtext.ast.ListItemChild
import com.company.richtext.parser.MarkdownParser

/**
 * Converts Markdown + `<@userId|userName>` tokens into plaintext for Twilio outbound SMS.
 */
fun stripMarkdownAndMentions(text: String): String {
    val document = MarkdownParser.parse(text)
    return stripDocument(document)
        .replace(Regex("""[ \t]+\n"""), "\n")
        .trim()
}

fun stripDocument(document: DocumentNode): String =
    document.children.joinToString("\n\n") { stripBlock(it) }

private fun stripBlock(node: BlockNode): String = when (node) {
    is BlockNode.Paragraph -> stripInline(node.children)
    is BlockNode.Heading -> stripInline(node.children)
    is BlockNode.CodeBlock -> node.value
    is BlockNode.ListBlock -> node.items.mapIndexed { index, item ->
        val marker = if (node.ordered) "${(node.start ?: 1) + index}. " else "- "
        item.children.joinToString("\n") { child ->
            when (child) {
                is ListItemChild.Paragraph -> marker + stripInline(child.children)
                is ListItemChild.NestedList -> stripBlock(
                    BlockNode.ListBlock(child.ordered, child.start, child.items),
                ).lineSequence().joinToString("\n") { "  $it" }
            }
        }
    }.joinToString("\n")
}

private fun stripInline(nodes: List<InlineNode>): String =
    nodes.joinToString("") { stripInlineNode(it) }

private fun stripInlineNode(node: InlineNode): String = when (node) {
    is InlineNode.Text -> node.value
    InlineNode.HardBreak -> "\n"
    is InlineNode.Mention -> "@${node.payload.userName}"
    is InlineNode.InlineCode -> node.value
    is InlineNode.Bold -> stripInline(node.children)
    is InlineNode.Italic -> stripInline(node.children)
    is InlineNode.Link -> {
        val label = stripInline(node.children)
        when {
            label.isEmpty() -> node.href
            label == node.href -> node.href
            else -> "$label (${node.href})"
        }
    }
}

File: packages/ios/Sources/RichTextiOS/Strip.swift
import Foundation

public enum TwilioText {
    /// Converts Markdown + `<@userId|userName>` tokens into plaintext for outbound SMS.
    public static func stripMarkdownAndMentions(_ text: String) -> String {
        let document = MarkdownParser.parse(text)
        return strip(document)
            .replacingOccurrences(of: "[ \\t]+\\n", with: "\n", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    public static func strip(_ document: DocumentNode) -> String {
        document.children.map(stripBlock).joined(separator: "\n\n")
    }

    private static func stripBlock(_ node: BlockNode) -> String {
        switch node {
        case .paragraph(let children):
            return stripInline(children)
        case .heading(_, let children):
            return stripInline(children)
        case .codeBlock(_, let value):
            return value
        case .list(let ordered, let start, let items):
            return items.enumerated().map { index, item in
                let marker = ordered ? "\((start ?? 1) + index). " : "- "
                return item.children.map { child in
                    switch child {
                    case .paragraph(let children):
                        return marker + stripInline(children)
                    case .list(let nestedOrdered, let nestedStart, let nestedItems):
                        return stripBlock(
                            .list(ordered: nestedOrdered, start: nestedStart, items: nestedItems)
                        )
                        .split(separator: "\n")
                        .map { "  \($0)" }
                        .joined(separator: "\n")
                    }
                }.joined(separator: "\n")
            }.joined(separator: "\n")
        }
    }

    private static func stripInline(_ nodes: [InlineNode]) -> String {
        nodes.map(stripInlineNode).joined()
    }

    private static func stripInlineNode(_ node: InlineNode) -> String {
        switch node {
        case .text(let value):
            return value
        case .hardBreak:
            return "\n"
        case .mention(let payload):
            return "@\(payload.userName)"
        case .inlineCode(let value):
            return value
        case .bold(let children), .italic(let children):
            return stripInline(children)
        case .link(let href, let children):
            let label = stripInline(children)
            if label.isEmpty { return href }
            if label == href { return href }
            return "\(label) (\(href))"
        }
    }
}

public func stripMarkdownAndMentions(_ text: String) -> String {
    TwilioText.stripMarkdownAndMentions(text)
}
